File: leetcode/0676-implement-magic-dictionary/MagicDictionary.go
package leetcode

type MagicDictionary struct {
	dictionary []string
}

func Constructor() MagicDictionary {
	return MagicDictionary{}
}

func (MagicDictionary *MagicDictionary) BuildDict(dictionary []string) {
	MagicDictionary.dictionary = dictionary
}

func (MagicDictionary *MagicDictionary) Search(searchWord string) bool {
	for _, word := range MagicDictionary.dictionary {
		if len(word) != len(searchWord) {
			continue
		}

		var diff = 0
		for i := 0; i < len(word); i++ {
			var letter = word[i]
			if letter != searchWord[i] {
				diff++
			}

			if diff > 1 {
				break
			}
		}

		if diff == 1 {
			return true
		}
	}

	return false
}

File: leetcode/0676-implement-magic-dictionary/MagicDictionary.ts
class MagicDictionary {
  private dictionary: string[];
  constructor() {}

  buildDict(dictionary: string[]): void {
    this.dictionary = dictionary;
  }

  search(searchWord: string): boolean {
    for (const word of this.dictionary) {
      if (word.length !== searchWord.length) {
        continue;
      }

      let diff = 0;
      for (let index = 0; index < word.length; index++) {
        const letter = word[index];
        if (searchWord[index] !== letter) {
          diff++;
        }

        if (diff > 1) {
          break;
        }
      }

      if (diff === 1) {
        return true;
      }
    }

    return false;
  }
}

export default MagicDictionary;
